app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,103 @@
+
+      private
+
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # among other details the subscription_id, resource group, and provider identity needed for authorization.
+      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
+      # xms_mirid claim and populate a representative hash with the appropriate fields.
+      def token_identity_from_claims
+        @token_identity = {
+          subscription_id: xms_mirid_hash["subscriptions"],
+          resource_group:  xms_mirid_hash["resourcegroups"]
+        }
+
+        # determines which Azure assigned identity is provided in annotations
+        # user-assigned-identity:
+        #   - validates that the correct provider (Microsoft.ManagedIdentity) for a user identity is defined in the
+        #     claim. If so, the 'user_assigned_identity' attribute will be added to the hash with the corresponding
+        #     value from xms_mirid claim
+        # system-assigned-identity:
+        #   - validates that the correct provider (Microsoft.Compute) for a system identity is defined in the
+        #     claim and its resource is one that we support. At current, we only support a virtualMachines resource.
+        #     If these conditions are met, a 'system_assigned_identity' attribute will be added to the hash with its
+        #     value being the oid field in the JWT token.
+        @logger.debug(Log::ExtractingIdentityForAuthentication.new("#{xms_mirid_hash["providers"]}/#{xms_mirid_hash.keys[-1]}"))
+        if xms_mirid_hash["providers"] == "Microsoft.ManagedIdentity"
+          @token_identity[:user_assigned_identity] = xms_mirid_hash["userAssignedIdentities"]
+          @token_identity[:resource_name]          = xms_mirid_hash["userAssignedIdentities"]
+        else
+          @token_identity[:system_assigned_identity] = @oid_token_field
+          @token_identity[:resource_name]            = @oid_token_field
+        end
+      end
+
+      def validate_xms_mirid_format
+        valid_xms_mirid_format = xms_mirid_hash.length == 4 && xms_mirid_hash.key?("subscriptions" && "resourcegroups" && "providers")
+        raise Err::ClaimInInvalidFormat unless valid_xms_mirid_format
+      end
+
+      # we expect the xms_mirid claim to be in the format of /subscriptions/<subscription-id>/...
+      # therefore, we are ignoring the first slash of the xms_mirid field and grouping the entries in key-value pairs.
+      # ultimately, transforming "/key1/value1/key2/value2" to {"key1" => "value1", "key2" => "value2"}
+      def xms_mirid_hash
+        @xms_mirid_hash ||= Hash[*@xms_mirid_token_field.split('/').drop(1)]
+      end
+
+      # validate the integrity of annotations against the xms_mirid object representation
+      def validate_application_identity
+        @logger.debug(Log::ValidatingApplicationIdentity.new(@token_identity[:resource_name]))
```

Comment:
Let's remove this log message and also `:resource_name` from the token_identity hash. I don't think it helps anyway. If an annotation is not validated successfully then an error with it's name will be reaised.